### ISSUE TITLE
Version Knative YAMLs in Kyma repository

### ISSUE THREAD

**Issue by sjanota** [kyma-project/kyma#1785]
`knative-installer` currently downloads Knative releases during image building. Those images must be versioned in kyma repository so we are bound to specific Knative version.

**Comment by sjanota** [comment]
Originally posted by @Abd4llA:

I am very concerned with choosing the option to copy and maintain our knative manifests and drag along a dependency as a chart. I think we should

Be very clear about our dependencies and document them properly "e.g. k8s version, istio version and knative version".
Add a helper script to install it for minikube for a better developer experience.
But completely maintaining a version of knative CRDs is a dangerous choice IMHO. Adds an overhead for maintaining the chart, the upgrades, and upgrading is gonna be very tricky. I would rather have:

Kyma released against a release version of knative. Have that be documented and changed with every release "e.g. Kyma 0.5 with Knative 0.1".
Kyma master working with knative latest.
Have Kyma master break the day knative dependencies break, this way the change was introduced yesterday in knative probably and we can figure it out easily, as opposed to hiding behind a copy of knative CRDs and finding out what changed over the last two months in order to upgrade to newer version.

**Comment by sjanota** [comment]
> 1. Be very clear about our dependencies and document them properly "e.g. k8s version, istio version and knative version".

Totally agree. This information is included for k8s and istio, and will be for KNative regardless the option we will choose.

> 2. Add a helper script to install it for minikube for a better developer experience.

While it makes Kyma's "core" (whatever it is) clear it puts on us the burden of maintaining Bash and Powershell scripts. I would rather make it a job so it's OS agnostic and can be developed in more civilised languages. Or we could have a CLI which could do that sort of things, but that's much broader topic.

> But completely maintaining a version of knative CRDs is a dangerous choice IMHO. Adds an overhead for maintaining the chart, the upgrades, and upgrading is gonna be very tricky. 

+💯 for the upgrades part

> 1. Kyma released against a release version of knative. Have that be documented and changed with every release "e.g. Kyma 0.5 with Knative 0.1".
> 2. Kyma master working with knative latest.
> 3. Have Kyma master break the day knative dependencies break, this way the change was introduced yesterday in knative probably and we can figure it out easily, as opposed to hiding behind a copy of knative CRDs and finding out what changed over the last two months in order to upgrade to newer version.

I can't agree. Kyma should be bound to specific version of KNative as we both have products in very early stage of development. We would put on ourselves unnecessary burden of catching up with KNative every time they do a breaking change. I'm a fan of frequent updates of dependencies, but I'd like to do them on my terms. When I have manpower to do them, not when KNative team decided to do release. If we bind ourselves to their latest KNative's release cycle is controlling our development plans.

That being said I'm against having more maintenance (of not our codebase) on our heads. The only reason I see why it's reasonable to make KNative a chart (now) is to be consistent with installer, which has very little value for me. 

What I see as a solution here is:
* Have a job which installs KNative and modifies it if necessary (maybe split in two steps, as with istio)
* When KNative introduces breaking changes in next release use new version and fix those modifications
* When they are ready with helm charts migrate the process to use them instead of YAMLs

Gain:
* Upgrades are easy until they have helm support
* The migration is done once (when they are ready with charts) instead of twice (now and migrating our chart to theirs).

I don't have opinion about storing raw release files in kyma repository. I think I'm leaning towards having them checked in, because: 
1. We need to have the version hardcoded somewhere
2. Files in their release shouldn't change

If we keep them raw upgrade means only downloading new file and that's acceptable "maintenance" for me. On the other hand we left far behind times when JARs were checked into VCS, so why charts should be :)

**Comment by Abd4llA** [comment]
> I would rather make it a job so it's OS agnostic

+1, moving it to a job was a good step. I didn't really have a preference for a script.

> I can't agree. Kyma should be bound to specific version of KNative as we both have products in very early stage of development. We would put on ourselves unnecessary burden of catching up with KNative every time they do a breaking change. I'm a fan of frequent updates of dependencies, but I'd like to do them on my terms. When I have manpower to do them, not when KNative team decided to do release. If we bind ourselves to their latest KNative's release cycle is controlling our development plans.

I agree. I think I was thinking of the ideal world where we are very involved in knative decision making then it'd be probably long expected upfront as you see with Pivotal guys, and we would be part of the decision of introducing breaking changes. So since we're not there, it would be good to pinpoint to a "released" version. And try to regularly update it with every new "release" but in a controlled manner. 

So +1 from me for no "latest".

> What I see as a solution here is:
> 
> Have a job which installs KNative and modifies it if necessary (maybe split in two steps, as with istio)
> When KNative introduces breaking changes in next release use new version and fix those modifications
> When they are ready with helm charts migrate the process to use them instead of YAMLs

+1 . That would be a good solution.

**Comment by sjanota** [comment]
Originally posted by @Abd4llA [here](https://github.com/kyma-project/kyma/pull/1803#issuecomment-441641604):
> I would rather make it a job so it's OS agnostic

+1, moving it to a job was a good step. I didn't really have a preference for a script.

> I can't agree. Kyma should be bound to specific version of KNative as we both have products in very early stage of development. We would put on ourselves unnecessary burden of catching up with KNative every time they do a breaking change. I'm a fan of frequent updates of dependencies, but I'd like to do them on my terms. When I have manpower to do them, not when KNative team decided to do release. If we bind ourselves to their latest KNative's release cycle is controlling our development plans.

I agree. I think I was thinking of the ideal world where we are very involved in knative decision making then it'd be probably long expected upfront as you see with Pivotal guys, and we would be part of the decision of introducing breaking changes. So since we're not there, it would be good to pinpoint to a "released" version. And try to regularly update it with every new "release" but in a controlled manner. 

So +1 from me for no "latest".

> What I see as a solution here is:
> 
> Have a job which installs KNative and modifies it if necessary (maybe split in two steps, as with istio)
> When KNative introduces breaking changes in next release use new version and fix those modifications
> When they are ready with helm charts migrate the process to use them instead of YAMLs

+1 . That would be a good solution.

**Comment by sjanota** [comment]
@Abd4llA @piotrmsc @pbochynski 
I did a follow up on this and tried to implement #1598 using helm chart. As I said there are some changes in KNative's YAMLs which are unavoidable. On the other hand, implementation of the switch to enable/disable KNative mode was a way easier in charts. It isn't pretty, but nor are the bash scripts.

So to summarise my impression:
 * version with chart is a way better to handle all of the customisations we have to do
 * version with yamls is a way better when it comes to upgrades

Other then that both approaches are more or less equal:
 * they both require us to version KNative release
 * they are both "ugly" in a way

Also having it as a chart has a hidden bonus of having the switch as single installer override. There's pretty much nothing that has to be done in bash/ps scripts except for creating additional ConfigMap.

**Comment by pbochynski** [comment]
I like the change #1803 very much. It is much more clean then previous scripts. Next step is just contact knative community and contribute the charts. This way we have win win situation.

**Comment by sjanota** [comment]
@pbochynski 
Maybe a review then? ;)

@Abd4llA 
Are you satisfied with the solution?

**Comment by piotrmsc** [comment]
I agree with @pbochynski with the contribution to knative, at least ask them if they're interested and then create initial PR with charts. It's a good point to contribute instead of waiting, and I know that from the experience with istio since 0.3... ;) 

Having a kubernetes job that will execute bash scripts to modify & apply yaml files does not sound good to me for a couple of reasons:

- it breaks our initiative to reduce bash scripts to the minimum
- applying changes for example regarding FAAS or eventing will be easier to set in the yaml within a chart rather then modify scripts
- versioning is easier to maintain (we cannot use latest version all the time)  
- applying changes for local environment only is easier to do in the helm chart
- with helm charts we can make use installer overrides instead of creating and mounting manually files to the container etc. Which is also nice from the upgrade phase perspective
- we do not have yet another component responsible for installing something in the kyma -  kyma installer is responsible for that
- and last but not least we are consistent across core by using helm (remember about our decision to integrate with knative)

I agree with "maintenance costs" but not entirely. @Abd4llA If you need further discussion regarding this topic with me or @pbochynski  let's move it offline (will be faster) and paste here additional notes worth mentioning.

**Comment by Abd4llA** [comment]
As @sjanota already said, each solution has pros and cons. So I'll remove by block and leave it up to the team who'll own maintaining knative to make decide on their own if they're willing to commit to maintaining it or not, now they've three point of views who're all outside the team. 

- My POV is not favoring the scripts, it is only about diverting from how knative release their components. @pbochynski Contributing the chart is a great idea! And I think it is a decent middle ground where the team might be encouraged to move forward with that approach given that they'll push forward in contributing the chart.

Again as I said, I leave it now to the team to commit to the solution they see best and willing to commit behind it.
